### PR TITLE
Resolve hostnames

### PR DESCRIPTION
**Pull request by dgalaktionov**
Resolves hostnames (such as `localhost` or `github.com`) for endpoint addresses. Gives preference to IPv4 over IPv6, although it works with both.

This is a small first contribution where my main goal was to be able to build the project and familiarize with the code. Feel free to point out any issues you may see!

### CODE REVIEW

**Comment by maximegmd** [comment]
Hello, thank you for the PR.
I have a few observations:

- The code guidelines are not followed (but that's not your fault, I have yet to write them down), in this instance it's the argument names, they need to be prefixed with an 'a' + 'c' if const + 'p' if pointer for example: void Test(const int* acpSomeValue); 

- The endpoint class should not be the one responsible for name resolution, we probably need a dedicated class for this "Resolver" that could be started asynchronously and contain a list of all the endpoints after resolution. On the top of my head I am thinking it could be a great use of std::async and std::future.

- The for loop used to set the hostname will loop until the last element is reached, could break the loop once you find the ipv4 entry you are looking for.

Cheers,
Max

**Comment by dgalaktionov** [comment]
Thank you for the quick response.

How about moving all the parsing (and resolving) logic to a `Resolver`? It would be constructed with a string and have a `get` method that would return a vector of `Endpoint` objects. That way only the `Resolver` has to know about `std::async` and `std::future`, and we would also make `Endpoint` simple and stupid. In the future, this `Resolver` could be useful not only for a client but also for a server, if we want to allow the user to choose their listening interface by setting an IP address (I think Minecraft does that).

We could further separate the logic into `IPv6Resolver`, `IPv4Resolver` and `HostnameResolver`, which could all be invoked by a `CascadeResolver`, but that's a question of how much overengineered OO design you can tolerate in your codebase 😁

**Comment by maximegmd** [comment]
I like the idea of using a Resolver and then maybe instead of a "get" method, add operator[] and iterators ?

The IPv4, IPv6 etc seems a bit too much ^^ We could template it though in case someone really only wants IPv4/v6 or both.

File: Code/tests/src/network.cpp [hunk 1]
```diff
@@ -1,331 +1,331 @@
-#include "catch.hpp"
```
Comment (maximegmd) [inline Code/tests/src/network.cpp]: Did you change the line endings by any chance ?

Comment (dgalaktionov) [reply Code/tests/src/network.cpp]: Good catch. I had autocrlf = input in my linux .gitconfig
Won't happen again!

**Comment by dgalaktionov** [comment]
Should be good now. I've removed the string constructor and parsing functions from `Endpoint`, as this kind of initialization must now me done through `Resolver`. I've tried to follow the guidelines this time too.

File: Code/network/src/Resolver.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+Resolver::Iterator Resolver::end() noexcept
+{
+    return Iterator(GetEndpoints().cend());
+}
+
+std::vector<Endpoint> & Resolver::GetEndpoints() noexcept
+{
+    if (m_futureEndpoints.valid()) {
```
Comment (maximegmd) [inline Code/network/src/Resolver.cpp]: { needs to be on a new line

File: Code/network/src/Resolver.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        m_endpoints.insert(std::end(m_endpoints), std::begin(newEndpoints), std::end(newEndpoints));
+    }
+
+    return m_endpoints;
+}
+
+void Resolver::Parse(const std::string & acAddress) noexcept
+{
+    uint16_t port = 0;
+    std::string strAddress(acAddress);
```
Comment (maximegmd) [inline Code/network/src/Resolver.cpp]: Might as well just pass the argument by value, this way we can get a move optimization

File: Code/network/src/Resolver.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+bool Resolver::IsEmpty() noexcept
+{
+    return GetSize() == 0;
+}
+
+Endpoint Resolver::GetEndpoint(const size_t & acIndex) noexcept
+{
+    return GetEndpoints().at(acIndex);
```
Comment (maximegmd) [inline Code/network/src/Resolver.cpp]: Don't use at(), it throws and I would love for the code to be exception free

**Review by maximegmd — CHANGES_REQUESTED** [review]

File: Code/network/src/Resolver.cpp (hunk 3 above)
Comment (dgalaktionov) [reply Code/network/src/Resolver.cpp]: Exactly my thought. I just copied it from your code and preferred not to touch it. Glad that you want it fixed 😄

File: Code/network/include/Resolver.h [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+public:
+    class Iterator {
+    public:
+        using difference_type = ptrdiff_t;
+        using value_type = Endpoint;
+        using pointer = const Endpoint*;
+        using reference = const Endpoint&;
+        using iterator_category = std::forward_iterator_tag;
+
+        Iterator() : m_iterator() {}
```
Comment (dgalaktionov) [inline Code/network/include/Resolver.h]: Should these braces go on new lines too?

Comment (maximegmd) [reply Code/network/include/Resolver.h]: No it's fine for empty stuff like that

File: Code/network/src/Resolver.cpp [hunk 6]
```diff
@@ -64,52 +65,51 @@ std::vector<Endpoint> & Resolver::GetEndpoints() noexcept
     return m_endpoints;
 }
 
-void Resolver::Parse(const std::string & acAddress) noexcept
+void Resolver::Parse(std::string aAdress) noexcept
```
Comment (maximegmd) [inline Code/network/src/Resolver.cpp]: address takes 2 d ^^

**Review by maximegmd — CHANGES_REQUESTED** [review]

**Pull request merged**